Use the builder pattern to create authenticators.

ggriffiniii opened the pull request:
Beyond simply moving to the builder pattern for intialization this has a
few other effects.

The DeviceFlow and InstalledFlow can no longer be used without an
associated Authenticator. This is becaus they no longer have any
publicly accessible constructor. All initialization goes through the
Authenticator. This also means that the flows are always initialized
with a clone of the hyper client used by the Authenticator.

The authenticator uses the builder pattern which allows omitting
optional fields. This means that if users simply want a default hyper
client, they don't need to create one explicitly. One will be created
automatically. If users want to specify a hyper client (maybe to allow
sharing a single client between different libraries) they can still do so
by using the hyper_client method on the builder. Additionally for both
AuthenticatorDelegate's and FlowDelegate's if the user does not specify
an override the default ones will be used.

The builders are now exposed publicly with the names of Authenicator,
InstalledFlow, and DeviceFlow. The structs that actually implement those
behaviors are now hidden and only expose the GetToken trait. This means
some methods that were previously publicly accessible are no longer
available, but the methods appeared to be implementation details that
probably shouldn't have been exposed anyway.

ggriffiniii committed:
Use the builder pattern to create authenticators.

Beyond simply moving to the builder pattern for intialization this has a
few other effects.

The DeviceFlow and InstalledFlow can no longer be used without an
associated Authenticator. This is becaus they no longer have any
publicly accessible constructor. All initialization goes through the
Authenticator. This also means that the flows are always initialized
with a clone of the hyper client used by the Authenticator.

The authenticator uses the builder pattern which allows omitting
optional fields. This means that if users simply want a default hyper
client, they don't need to create one explicitly. One will be created
automatically. If users want to specify a hyper client (maybe to allow
sharing a single client between different libraries) they can still do so
by using the hyper_client method on the builder. Additionally for both
AuthenticatorDelegate's and FlowDelegate's if the user does not specify
an override the default ones will be used.

The builders are now exposed publicly with the names of Authenicator,
InstalledFlow, and DeviceFlow. The structs that actually implement those
behaviors are now hidden and only expose the GetToken trait. This means
some methods that were previously publicly accessible are no longer
available, but the methods appeared to be implementation details that
probably shouldn't have been exposed anyway.

dermesser commented:
Sorry for the late response.

The change itself seems fine to me, but honestly I don't think that the various `new()` methods are so complicated as to warrant this change for now. Simply, because it adds more layers when reading documentation, as you now have the type and an `...Impl` type, and various methods etc.

Or do you think that the current number of constructor arguments is already too bad?

ggriffiniii commented:
I guess we have a difference of opinion on which is simpler to read and understand, and that's fine this is your codebase and you should feel free to accept or reject any changes as you see fit. I will accept your decision either way. I feel like the changes to the examples demonstrate why I prefer this approach so to highlight one of the examples:

This is without the builder pattern

```rust
    let https = HttpsConnector::new(1);
    let client = Client::builder()
        .keep_alive(false)
        .build::<_, hyper::Body>(https);
    let ad = yup_oauth2::DefaultFlowDelegate;
    let inf = InstalledFlow::new(
        client.clone(),
        ad,
        secret,
        yup_oauth2::InstalledFlowReturnMethod::HTTPRedirect(8081),
    );
    let mut auth = Authenticator::new_disk(
        client,
        inf,
        yup_oauth2::DefaultAuthenticatorDelegate,
        "tokencache.json",
    )
    .unwrap();
```

this is with the builder pattern

```rust
    let mut auth = Authenticator::new(InstalledFlow::new(
        secret,
        yup_oauth2::InstalledFlowReturnMethod::HTTPRedirect(8081),
    ))
    .persist_tokens_to_disk("tokencache.json")
    .build()
    .unwrap();
```

These do the exact same thing, but previously there were a lot of details that required thinking about. What hyper client should I use, or there's a note to disable keep-alives for some reason, better remember to do that. Let's also clone the same hyper client between the authenticator and the flow so they both can share the same connection pool. What flow delegate do I want? oh, there's a default lets start with that. Now there's a authenticator delegate? Let's use that default. All those questions have sane defaults when using the builder pattern so there's less upfront boilerplate to setting up a correctly working authenticator (while keeping the existing functionality for the users that need it). Additional benefits include enforcing in your API what comments recommend today, namely that a InstalledFlow and DeviceFlow should not be used without a corresponding Authenticator. In my mind that simply removes another potential footgun for users since it's not possible to create those flows directly anymore.

In the end a lot of this is just personal preference, so feel free to do what you think is best. This was merely a suggestion.

dermesser commented:
I agree, the setup of HTTP client, delegates, etc. is something I hadn't thought of. And the new style is obviously better for usage. As maintainer, it's easy to overlook these kinds of issues as long as it works :)

While you're at it; would you mind changing the names of the builder types (e.g. `Authenticator`) to say `...Builder` (`AuthenticatorBuilder`), and make `AuthenticatorImpl` to `Authenticator` again? When I first looked at the code, it caused a bit of confusion on my side, as I would usually expect some `XyImpl` to have a corresponding `Xy` trait, for example. If the type name says `Builder`, however, it is clear immediately what to expect from using it.

ggriffiniii committed:
Change the name of TokenGetterBuilder to AuthFlow.

I believe AuthFlow more succinctly describes the purpose of the type to
users reading documentation.

ggriffiniii commented:
I can change the names to whatever you want, but I wanted to point out that the builders are now the only types users actually see when looking at documentation. AuthenticatorImpl isn't even public, DeviceFlowImpl and InstalledFlowImpl are technically public but are scoped within a private module and the user should not ever see them. Authenticator::build returns an unnamed 'impl GetToken' and that's the only thing the user ends up interacting with. I didn't see value in adding a Builder suffix to each of the builder structs because as far as the public API is concerned there's nothing to confuse it with. Let me know if you would still prefer renaming the types with a Builder suffix.
